src/core/task/tools/handlers/WriteToFileToolHandler.ts

```diff
@@ -445,7 +285,142 @@ export class WriteToFileToolHandler implements IFullyManagedTool {
+			const errorResponse = formatResponse.toolError(formatResponse.clineIgnoreError(relPath))
+			ToolResultUtils.pushToolResult(
+				errorResponse,
+				block,
+				config.taskState.userMessageContent,
+				ToolDisplayUtils.getToolDescription,
+				config.api,
+				() => {
+					config.taskState.didAlreadyUseTool = true
+				},
+				config.coordinator,
+			)
+			return
 		}
+
+		// Check if file exists to determine the correct UI message
+		let fileExists: boolean
+		if (config.services.diffViewProvider.editType !== undefined) {
+			fileExists = config.services.diffViewProvider.editType === "modify"
+		} else {
+			const absolutePath = path.resolve(config.cwd, relPath)
+			fileExists = await fileExistsAtPath(absolutePath)
+			config.services.diffViewProvider.editType = fileExists ? "modify" : "create"
+		}
+
+		// Construct newContent from diff
+		let newContent: string
+		newContent = "" // default to original content if not editing
+
+		if (diff) {
+			// Handle replace_in_file with diff construction
+			if (!config.api.getModel().id.includes("claude")) {
+				// deepseek models tend to use unescaped html entities in diffs
+				diff = fixModelHtmlEscaping(diff)
+				diff = removeInvalidChars(diff)
+			}
+
+			// open the editor if not done already.  This is to fix diff error when model provides correct search-replace text but Cline throws error
+			// because file is not open.
+			if (!config.services.diffViewProvider.isEditing) {
+				await config.services.diffViewProvider.open(relPath)
+			}
+
+			try {
+				newContent = await constructNewFileContent(
+					diff,
+					config.services.diffViewProvider.originalContent || "",
+					!block.partial, // Pass the partial flag correctly
+				)
+			} catch (error) {
+				// Full original behavior - comprehensive error handling even for partial blocks
+				await config.callbacks.say("diff_error", relPath)
+
+				// Extract error type from error message if possible
+				const errorType =
+					error instanceof Error && error.message.includes("does not match anything")
+						? "search_not_found"
+						: "other_diff_error"
+
+				// Add telemetry for diff edit failure
+				telemetryService.captureDiffEditFailure(config.ulid, config.api.getModel().id, errorType)
+
+				// Push tool result with detailed error using existing utilities
+				const errorResponse = formatResponse.toolError(
+					`${(error as Error)?.message}\n\n` +
+						formatResponse.diffError(relPath, config.services.diffViewProvider.originalContent),
+				)
+				ToolResultUtils.pushToolResult(
+					errorResponse,
+					block,
+					config.taskState.userMessageContent,
+					ToolDisplayUtils.getToolDescription,
+					config.api,
+					() => {
+						config.taskState.didAlreadyUseTool = true
+					},
+					config.coordinator,
+				)
+
+				// Revert changes and reset diff view
+				await config.services.diffViewProvider.revertChanges()
+				await config.services.diffViewProvider.reset()
+
```

Comment:
@saoudrizwan in the previous code we had a checkpoint [here](https://github.com/cline/cline/blob/bd526cb2d2244efab79bb16942439d5edb38c782/src/core/task/ToolExecutor.ts#L418C8-L418C31) so we are the missing the checkpoint when a diff edit fails after some streaming. 

<img width="446" height="520" alt="image" src="https://github.com/user-attachments/assets/6d672702-032c-40ce-ad2f-e49683386979" />


I think in some ways that actually makes sense because if the diff edit fails then the checkpoint is frivolous but I wanted to point out the lack of parity.

Reply:
Good catch, and I agree it's unnecessary 👍